payment/src/main/java/net/optile/network/basic/BasicNetworkService.java

```diff
@@ -90,41 +102,69 @@ public void onRedirectCanceled() {
-        Interaction interaction = result.getInteraction();
-        String code = interaction.getCode();
-        if (!InteractionCode.PROCEED.equals(code)) {
-            presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_CANCELED, new PaymentResult(result));
+    public void onOperationSuccess(OperationResult operationResult) {
+        Interaction interaction = operationResult.getInteraction();
+        PaymentResult result = new PaymentResult(operationResult);
+        
+        if (!InteractionCode.PROCEED.equals(interaction.getCode())) {
+            presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_CANCELED, result);
             return;
         }
-        Redirect redirect = result.getRedirect();
-        if (redirect != null) {
-            switch (redirect.getType()) {
-                case RedirectType.PROVIDER:
-                case RedirectType.HANDLER3DS2:
-                    redirectPayment(redirect);
-                    return;
-            }
+        if (operationResult.getRedirect() != null) {
+            handleRedirect(operationResult);
+            return;
         }
-        presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_OK, new PaymentResult(result));
+        presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_OK, result);
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void onOperationError(Throwable cause) {
-        PaymentResult result = PaymentResult.fromThrowable(cause);
-        int status = result.hasError() ? PaymentUI.RESULT_CODE_ERROR : PaymentUI.RESULT_CODE_CANCELED;
-        presenter.onPreparePaymentResult(status, result);
+        PaymentError error = PaymentError.fromThrowable(cause);
+        handleProcessPaymentError(error);
+    }
+
+    private void handleRedirect(OperationResult operationResult) {
+        Redirect redirect = operationResult.getRedirect();
+        switch (redirect.getType()) {
+            case RedirectType.PROVIDER:
+            case RedirectType.HANDLER3DS2:
+                try {
+                    presenter.redirectPayment(redirect);
+                } catch (PaymentException e) {
+                    handleProcessPaymentError(e.error);
+                }
+                break;
+            default:
+                PaymentResult result = new PaymentResult(operationResult);
+                presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_OK, result);
+        }
+    }
+
+    private void handleProcessPaymentError(PaymentError error) {
+        ErrorInfo errorInfo = error.getErrorInfo();
+        PaymentResult result;
+
+        if (errorInfo != null) {
+            result = new PaymentResult(errorInfo.getInteraction(), error);
+        } else {
+            result = PaymentResult.fromPaymentError(error);
+            result.getInteraction().setCode(getErrorInteractionCode(operation));
+        }
+        presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_CANCELED, result);
```

Comment:
Still not very nice, but a bit more separation of concerns:
```suggestion
    private void handleProcessPaymentError(PaymentError error) {
        presenter.onProcessPaymentResult(PaymentUI.RESULT_CODE_CANCELED, toPaymentResult(error));
    }

    private PaymentResult toPaymentResult(PaymentError error) {
        ErrorInfo errorInfo = error.getErrorInfo();
        if (errorInfo != null) {
            return new PaymentResult(errorInfo.getInteraction(), error);
        }

        PaymentResult result = PaymentResult.fromPaymentError(error);
        result.getInteraction().setCode(getErrorInteractionCode(operation));
        return result;
```
Especially the `getInteraction().setCode(...)` bothers me - can you introduce a variant of `PaymentResult.fromPaymentError` that takes `Interaction` as the second parameter?

Reply:
Updated the BasicNetworkService and removed the static method to create a PaymentResult.